pkg/jobmanager/job.go

```diff
@@ -74,45 +88,129 @@ func (j *Job) ExitCode() int {
 	return c
 }
 
-// wait for the underlying process. Broadcast the end via waitChan
-// and close the given resources.
-func (j *Job) wait() {
-	if err := j.cmd.Wait(); err != nil {
-		slog.Debug("Process Wait ended with error", "error", err)
+// closeGroup freeze/kills the group then checks if it is still in use.
+// Returns true when no more process are attached to the group.
+func (j *Job) closeGroup() (bool, error) {
+	// Freeze the cgroup for good measure.
+	if err := os.WriteFile(filepath.Join(j.cgroupPath, "cgroup.freeze"), []byte("1"), 0); err != nil {
+		return false, fmt.Errorf("freeze group: %w", err)
+	}
+	// Kill all processes in the cgroup.
+	if err := os.WriteFile(filepath.Join(j.cgroupPath, "cgroup.kill"), []byte("1"), 0); err != nil {
+		return false, fmt.Errorf("kill group: %w", err)
+	}
+
+	// Assert that the group is empty.
+	buf, err := os.ReadFile(filepath.Join(j.cgroupPath, "cgroup.procs"))
+	if err != nil {
+		return false, fmt.Errorf("lookup group procs: %w", err)
 	}
 
+	return len(buf) == 0, nil
+}
+
+func (j *Job) close() {
 	j.mu.Lock()
-	defer j.mu.Unlock()
 	if j.status != pb.JobStatus_JOB_STATUS_STOPPED {
 		j.status = pb.JobStatus_JOB_STATUS_EXITED
 	}
-	j.exitCode = j.cmd.ProcessState.ExitCode()
+	if j.cmd.ProcessState != nil {
+		j.exitCode = j.cmd.ProcessState.ExitCode()
+	}
 	close(j.waitChan)
-	if err := j.broadcaster.Close(); err != nil {
+	if e1 := j.broadcaster.Close(); e1 != nil {
 		// Best effort.
-		slog.Error("Broadcaster closed with error.", "error", err)
+		slog.Error("Broadcaster closed with error.", "error", e1)
 	}
+	j.mu.Unlock()
+
+	// NOTE: cgroupPath is immutable and set at start before being shared, can
+	// safely be used without lock.
+	logger := slog.With("job_id", j.ID.String(), "cgroup_path", j.cgroupPath)
+
+	// Wait for ~1 second (arbitrary) for the cgroup to be empty.
+	const tickerInterval = 10 * time.Millisecond
+	const tickerAttempts = 100
+	ticker := time.NewTicker(tickerInterval)
+	defer ticker.Stop()
+	for range tickerAttempts {
+		empty, err := j.closeGroup()
+		if err != nil {
+			// Best effort.
+			logger.Error("Error closing group.", "error", err)
+			break
+		}
+		if empty {
+			if err := os.Remove(j.cgroupPath); err != nil {
+				// If cleanup failed, go over again freeze/kill/assert/cleanup.
+				// If it happens it likely means something violated the cgroup single writer principle.
+				logger.Error("Error removing cgroup on job close.", "error", err)
```

Comment:
I wonder if it's worth to check if the error is EBUSY and only retry if it is.

Reply:
Done.